subprojects/model-core/src/main/java/org/gradle/internal/properties/annotations/AbstractTypeMetadataWalker.java

```diff
@@ -0,0 +1,208 @@
+        visitor.visitRoot(typeMetadata, root);
+        walkChildren(root, typeMetadata, null, visitor, nestedNodeToQualifiedNameMapFactory.get());
+    }
+
+    private void walk(T node, String qualifiedName, PropertyMetadata propertyMetadata, NodeMetadataVisitor<T> visitor, Map<T, String> nestedNodesWalkedOnPath) {
+        Class<?> nodeType = resolveType(node);
+        TypeMetadata typeMetadata = typeMetadataStore.getTypeMetadata(nodeType);
+        if (Provider.class.isAssignableFrom(nodeType)) {
+            handleProvider(node, child -> walk(child, qualifiedName, propertyMetadata, visitor, nestedNodesWalkedOnPath));
+        } else if (Map.class.isAssignableFrom(nodeType) && !typeMetadata.hasAnnotatedProperties()) {
+            handleMap(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), propertyMetadata, visitor, nestedNodesWalkedOnPath));
+        } else if (Iterable.class.isAssignableFrom(nodeType) && !typeMetadata.hasAnnotatedProperties()) {
+            handleIterable(node, (name, child) -> walk(child, getQualifiedName(qualifiedName, name), propertyMetadata, visitor, nestedNodesWalkedOnPath));
+        } else {
+            handleNested(node, typeMetadata, qualifiedName, propertyMetadata, visitor, nestedNodesWalkedOnPath);
+        }
+    }
+
+    private void walkChildren(T node, TypeMetadata typeMetadata, @Nullable String parentQualifiedName, NodeMetadataVisitor<T> visitor, Map<T, String> nestedNodesOnPath) {
+        typeMetadata.getPropertiesMetadata().forEach(propertyMetadata -> {
+            String childQualifiedName = getQualifiedName(parentQualifiedName, propertyMetadata.getPropertyName());
+            if (propertyMetadata.getPropertyType() == nestedAnnotation) {
+                Optional<T> childOptional = getChild(node, propertyMetadata);
+                childOptional.ifPresent(child -> walk(child, childQualifiedName, propertyMetadata, visitor, nestedNodesOnPath));
+            } else {
+                visitor.visitLeaf(childQualifiedName, propertyMetadata, () -> getChild(node, propertyMetadata).orElse(null));
+            }
+        });
+    }
+
+    private void handleNested(T node, TypeMetadata typeMetadata, String qualifiedName, PropertyMetadata propertyMetadata, NodeMetadataVisitor<T> visitor, Map<T, String> nestedNodesOnPath) {
+        String firstOccurrenceQualifiedName = nestedNodesOnPath.putIfAbsent(node, qualifiedName);
+        if (firstOccurrenceQualifiedName != null) {
+            onNestedNodeCycle(firstOccurrenceQualifiedName, qualifiedName);
+            return;
+        }
+
+        visitor.visitNested(typeMetadata, qualifiedName, propertyMetadata, node);
+        walkChildren(node, typeMetadata, qualifiedName, visitor, nestedNodesOnPath);
+        nestedNodesOnPath.remove(node);
+    }
+
+    abstract void onNestedNodeCycle(@Nullable String firstOccurrenceQualifiedName, String secondOccurrenceQualifiedName);
+
+    abstract void handleProvider(T node, Consumer<T> handler);
```

Comment:
❓ Why aren't we using `protected` here like we do for all the other `abstract` methods?

Reply:
Overlooked, should be fixed now.